RecoMTD/DetLayers/src/MTDRingForwardDoubleLayer.cc

```diff
@@ -0,0 +1,235 @@
+/** \file
+ *
+ *  \author L. Gray
+ */
+
+#include <RecoMTD/DetLayers/interface/MTDRingForwardDoubleLayer.h>
+#include <RecoMTD/DetLayers/interface/MTDDetRing.h>
+#include <Geometry/CommonDetUnit/interface/GeomDet.h>
+#include <DataFormats/GeometrySurface/interface/SimpleDiskBounds.h>
+#include <TrackingTools/GeomPropagators/interface/Propagator.h>
+#include <TrackingTools/DetLayers/interface/MeasurementEstimator.h>
+
+#include <FWCore/MessageLogger/interface/MessageLogger.h>
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+MTDRingForwardDoubleLayer::MTDRingForwardDoubleLayer(const vector<const ForwardDetRing*>& frontRings,
+                                       const vector<const ForwardDetRing*>& backRings) :
+  RingedForwardLayer(true),
+  theFrontLayer(frontRings),
+  theBackLayer(backRings),
+  theRings(frontRings), // add back later
+  theComponents(),
+  theBasicComponents()
+{
+
+  const std::string metname = "MTD|RecoMTD|RecoMTDDetLayers|MTDRingForwardDoubleLayer";
+
+  theRings.insert(theRings.end(), backRings.begin(), backRings.end());
+  theComponents = std::vector <const GeometricSearchDet*>(theRings.begin(), theRings.end());
+
+  // Cache chamber pointers (the basic components_)
+  // and find extension in R and Z
+  for (vector<const ForwardDetRing*>::const_iterator it=theRings.begin();
+       it!=theRings.end(); it++) {
+    vector<const GeomDet*> tmp2 = (*it)->basicComponents();
+    theBasicComponents.insert(theBasicComponents.end(),tmp2.begin(),tmp2.end());
+  }  
+  
+  setSurface(computeSurface());
+   
+  LogTrace(metname) << "Constructing MTDRingForwardDoubleLayer: "
+                    << basicComponents().size() << " Dets " 
+                    << theRings.size() << " Rings "
+                    << " Z: " << specificSurface().position().z()
+                    << " R1: " << specificSurface().innerRadius()
+                    << " R2: " << specificSurface().outerRadius();
+
+  selfTest();
+}
+
+
+BoundDisk * MTDRingForwardDoubleLayer::computeSurface() 
+{
+  const BoundDisk & frontDisk = theFrontLayer.specificSurface();
+  const BoundDisk & backDisk  = theBackLayer.specificSurface();
+
+  float rmin = min( frontDisk.innerRadius(), backDisk.innerRadius() );
+  float rmax = max( frontDisk.outerRadius(), backDisk.outerRadius() );
+  float zmin = frontDisk.position().z();
+  float halfThickness = frontDisk.bounds().thickness()/2.;
+  zmin = (zmin > 0) ? zmin-halfThickness : zmin+halfThickness;
+  float zmax = backDisk.position().z();
+  halfThickness = backDisk.bounds().thickness()/2.;
+  zmax = (zmax > 0) ? zmax+halfThickness : zmax-halfThickness;
+  float zPos = (zmax+zmin)/2.;
+  PositionType pos(0.,0.,zPos);
+  RotationType rot;
+
+  return new BoundDisk( pos, rot,
```

Comment:
This is handled by a ReferenceCountingPointer in the base class.

Reply:
https://github.com/cms-sw/cmssw/blob/master/TrackingTools/DetLayers/src/ForwardDetLayer.cc#L16

Which gives ownership to the ```ReferenceCountingPointer<BoundDisk>``` at https://github.com/cms-sw/cmssw/blob/master/TrackingTools/DetLayers/interface/ForwardDetLayer.h#L61

Modernizing this would be for another PR since it touches all tracking geometries in CMSSW.